lib/spack/spack/solver/concretize.lp

```diff
@@ -689,15 +696,28 @@ requirement_group_satisfied(node(ID, Package), X) :-
 % flags if their only source is from a requirement. This is overly-specific
 % and should use a more-generic approach like in https://github.com/spack/spack/pull/37180
 
-{ attr("node_flag", node(ID, A1), A2, A3) } :-
-  requirement_group_member(Y, Package, X),
-  activate_requirement(node(ID, Package), X),
-  imposed_constraint(Y,"node_flag_set", A1, A2, A3).
-
-{ attr("node_flag_source", node(ID, A1), A2, node(ID, A3)) } :-
-  requirement_group_member(Y, Package, X),
-  activate_requirement(node(ID, Package), X),
-  imposed_constraint(Y,"node_flag_source", A1, A2, A3).
+% None of the following 3 are needed for `virtual-w:require: w cflags="-O1"`
+% (1) and (3) are needed for `y:require: cflags="-g"`
```

Comment:
Regarding this and your point

> That said, the mpi case works also with this patch - I'll try to understand why since the names of the virtual "mpi" and of the package "openmpi" don't match the pattern above 

I will note that `emit_facts_from_requirement_rules` has differing behavior for virtuals vs. providers:

```
For y:

pkg_fact("y",trigger_id(0)).
pkg_fact("y",trigger_msg("y cflags=\"-g\"")).
condition_requirement(0,"node","y").
condition_requirement(0,"node_flag","y","cflags","-g").
condition_requirement(0,"node_flag_source","y","cflags","y").

For w:

pkg_fact("virtual-w",trigger_id(0)).
pkg_fact("virtual-w",trigger_msg("virtual-w")).
condition_requirement(0,"virtual_node","virtual-w").
```

the rule is easily satisfied for virtuals, but for non-virtuals the "condition" is equal to the constraint, I think that's to support groups, but IMO it ought to be refactored into something sensible.

Reply:
Update: none of these are required if using:

```

{ impose(EffectID, node(ID, Package)) } :-
  pkg_fact(Package, condition_effect(Y, EffectID)),
  requirement_group_member(Y, Package, X),
  activate_requirement(node(ID, Package), X),
  requirement_group(Package, X).
```

(added above)